Mal Auto Updates

toyhammered opened the pull request:
**Feature:** Kitsu will update your mal list so it is in sync with Kitsu at all times

~~**[x]** - when a user updates a library entry it will update mal list also~~
~~**[x]** - when a user initially syncs their account (inputting username/password) we will completely sync    up their mal list to be the same as kitsu list~~
  ~~- [x] will create/update all entries~~
    - [Future] will delete any entries on mal list and not on kitsu list
~~**[x]** - write behavior tests for `get, delete, post, put` methods~~
~~**[x]** - make sure this works with anime and manga~~
  ~~- **[x]** anime~~
  ~~- **[x]** manga~~
~~**[x]** - Encrypt linked_profile tokens (all of them)~~

toyhammered committed:
Update MAL list if LinkedProfile exists.
  - added post update routine to LibraryEntry
  - added http abstraction for Typhoeus
  - set up base structure for tests
    - tests still need to be implemented for anime/manga
  - check if manga works

toyhammered committed:
remove request.method and add after_destroy method

toyhammered committed:
add tests for http get

toyhammered committed:
move all newly created files to service

toyhammered committed:
add initial library check to sync mal

toyhammered committed:
add after save for linked profile

toyhammered committed:
fix rubocop errors. WIP tests

toyhammered committed:
fix more rubocop errors

toyhammered commented on app/workers/my_anime_list_list_worker.rb:
```diff
@@ -0,0 +1,16 @@
+class MyAnimeListListWorker
+  include Sidekiq::Worker
+
+  def perform(user_id)
```
hey.. if you read like 9 lines down in the private method IT IS RIGHT THERE?!?!! 🖕

NuckChorris replied:
You do realize that the private method can't access the parameter of this, right?  You need to pass it down or set it as an instance variable or property accessor

toyhammered replied:
oh I am such a dimwit sometimes...............

toyhammered replied:
forgetting ruby basics 101?

toyhammered committed:
give access to user_id in kitsu_library

toyhammered committed:
add test for #get method

toyhammered committed:
add behavior/unit testing.

toyhammered committed:
seperate anime/manga lookups

toyhammered committed:
Add token encryption
  - installed attr_encrypted gem
  - changed linked_profile columns
  - added static name to lp factory
  - need to figure out why Webmock is retarded
    - failing the before_save to verify credentials
    - cannot stub request
    - seriously, Webmock is retarded...

toyhammered commented:
@NuckChorris I have no idea why Webmock is not working for my before_save verify credentials (inside `linked_profile.rb`). I had Sven _attempting_ to help me, I literally copied/pasted the request webmock gave us and it still failed. Any advice on how to fix this?

Also feel free to review what we have done (I will fix the hound errors at some point before the merge)

toyhammered committed:
change before save to validation, and will fix some rubocop errors

toyhammered commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+            post(
+              "animelist/#{media_type}",
+              profile,
+              {
```
I actually don't understand what this means, it won't work without the `{}`.

NuckChorris replied:
really?  It _should_ work, I'm pretty sure that's valid in Ruby

toyhammered replied:
```
post(
              "animelist/#{media_type}",
              profile,
              anime_id: mal_media_id,
              status: format_status(le.status),
              episodes: le.progress,
              score: format_score(le.rating)
            )
```
looks really strange...

NuckChorris replied:
I think it's more that the indentation you've used (putting the method name on its own line) is really strange.  Keyword parameters and Hash params without curlies are standard procedure in Ruby

toyhammered commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+      )
+    else
+      # find the anime or manga
+      get("#{media_type}/#{mal_media_id}#{MINE}", profile) do |response|
```
I would get this error... Thanks hound...

NuckChorris replied:
You can probably split this up a little better, perhaps move the `put` stuff below into methods like `update_animelist(mal_media_id, le)` so that this is more clear what's going on.  Otherwise it's kind of difficult to visually track how indented you are at any given point

NuckChorris commented on .env:
```diff
@@ -23,3 +23,5 @@ STREAM_API_SECRET=uxupxkuw5ckrd2fe9qz88zne4qagwmdm9yaegbsm8xuymqja6ff8jr678636v3
 TWITTER_CONSUMER_KEY=some_key
 TWITTER_CONSUMER_SECRET=some_secret
+
+LP_TOKEN_KEY=\xBB_\xAE4\xBB\xDFG\x97\xDA\xAF\xEF\v\xF8\x843\x8B^\x13\xD1\xF5\xB6\x92Y\rT$\xA9\xB5i4\\y
```
Could you use `Base64.decode64` to read the variable, so we can base64-encode this and not put binary into an environment variable?

Also I'd just call it `ATTR_ENCRYPT_KEY` so we can use it generically

toyhammered replied:
not sure what you mean here.

NuckChorris commented on app/models/library_entry.rb:
```diff
@@ -104,6 +104,10 @@ def activity
     MediaActivityService.new(self)
   end
 
+  def linked_profile_exists?
```
This method name doesn't reveal that we're seeking a LinkedProfile _specifically for MAL_

It really ought to have a more specific name than this, perhaps `sync_to_mal?`

NuckChorris commented on app/models/library_entry.rb:
```diff
@@ -132,5 +136,20 @@ def activity
+
+    # Sync MAL updates if linked profile exists
+    if linked_profile_exists?
+      return unless media_type.in? %w[anime manga]
```
This will never be true, because the column is capitalized

`%w[Anime Manga]`

toyhammered replied:
ahh yes polymorphic rules...

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -31,6 +31,35 @@ class LinkedProfile < ApplicationRecord
+    #   errors.add(:token, "#{response.code}: #{response.body}")
+    # end
+
+    # true
```
Why is this all commented out?

toyhammered replied:
Sven and I were having Webmock issues with testing it, like it flat out would not work (so that code should be valid, was just having testing issues)

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+  ATARASHII_API_HOST = 'https://hbv3-mal-api.herokuapp.com/2.1/'.freeze
+  MINE = '?mine=1'.freeze
+
+  attr_reader :le, :method
```
I'd prefer if you used `library_entry` or `entry` for this.  `le` is not very descriptive, so should be limited to small contexts of just a few lines at most

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -31,6 +31,35 @@ class LinkedProfile < ApplicationRecord
+  end
+
+  after_save do
+    if url == 'myanimelist'
```
Why are you using `myanimelist` as the URL?  That's not a URL, that's a site name, why not just verify whether the LinkedSite is MAL?

toyhammered replied:
yes yes yes... Oh god I screwed up here...

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+        if media_type == 'anime'
+          if response['watched_status']
+            put(
+              "animelist/#{media_type}/#{mal_media_id}",
```
At this point you know it's an animelist, you don't need `media_type`, you can just directly cram `animelist/anime` in there and avoid the interpolation

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+        "#{animelist}/#{mal_media_type}/#{mal_media_id}",
+        profile
+      )
+    else
```
It's better if you use `when 'create/update'`, just in the off-chance we need to add more methods down the road.  Plus it's more clear what's going on

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+      user_id: le.user_id,
+      url: 'myanimelist'
+    )
+    mal_media_id = mal_media.external_id
```
These four lovelies can probably be moved to their own methods

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+    case method
+    when 'delete'
+      delete(
+        "#{animelist}/#{mal_media_type}/#{mal_media_id}",
```
>`mal_media_type` is undefined
>`animelist` is undefined

You can probably express this as:

```ruby
delete("#{media_type}list/#{media_type}/#{mal_media_id}", profile)
```

toyhammered replied:
yeah... I am so stupid.

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+                score: format_score(le.rating),
+                reread_count: le.reconsume_count
+              }
+            )
```
All the things which applied to the `put` for anime above also apply here

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+                chapters: le.progress,
+                score: format_score(le.rating)
+              }
+            )
```
And also here

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+  def format_score(score)
+    return nil if score.nil?
+
+    (score * 2).floor
```
Might be cleaner to just say

```ruby
def format_score(score)
  (score * 2).to_i if score
end
```

There's no reason to floor it, all our scores are .0 or .5, which when multiplied by two guarantees a .0

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+    request_status(request) do |response|
+      # doesn't need to do anything if success
+    end
+    request.run
```
I'd personally recommend just using Typheous in synchronous mode. If you call `Typhoeus::Request.new('https://example.com').run`, it returns a `Typheous::Response`.

You can call the `response_body`, `response_headers`, or `response_code` methods on the `Typheous::Response` to get info about the response

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+        $stderr.puts('HTTP request failed: ' + response.code.to_s)
+      end
+    end
+  end
```
It's probably better to raise an exception when we get a non-200 response so that it gets caught by Sentry and emailed to me ;)

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,203 @@
+        # to the parent request under my_anime_list.rb
+
+        # delete will not have a body so need try catch
+        yield response&.body
```
I don't think any of this is really necessary (except the basic concept of handling error cases, which is good!), since `post` and the other methods you've created will return the `Typheous::Response` due to ending in `request.run`... you can get the body of the response just by calling `post(things).response_body`

toyhammered committed:
Fix all issues from code review

toyhammered committed:
change decode to encode

toyhammered committed:
add gemfile.lock file

toyhammered committed:
add longer key, fix rubocop error

NuckChorris commented on app/models/library_entry.rb:
```diff
@@ -104,6 +104,14 @@ def activity
 
+  def sync_to_mal?
+    # is it better to take present? out and pass
+    # this variable into the worker also?
```
No, it makes more sense to handle it here, so it won't create a bunch of async jobs which do nothing

NuckChorris commented on app/models/library_entry.rb:
```diff
@@ -134,5 +142,20 @@ def activity
+
+    # Sync MAL updates if linked profile exists
+    if sync_to_mal?
+      return unless media_type.in? %w[Anime Manga]
```
Now that this is no longer named `linked_profile?`, maybe move this check into `sync_to_mal?`

Then it's just

```ruby
MyAnimeListSyncWorker.perform_async(self, 'create/update') if sync_to_mal?
```

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -31,6 +31,35 @@ class LinkedProfile < ApplicationRecord
+    User.find(user_id).linked_profiles.where(
+      linked_site: LinkedSite.find_by(name: 'myanimelist')
+    ).present?
+  end
```
This seems a bit... roundabout

Couldn't you just do `linked_site.name == 'myanimelist' && token.present?`

NuckChorris commented on app/models/linked_profile.rb:
```diff
@@ -31,6 +31,35 @@ class LinkedProfile < ApplicationRecord
   validates_presence_of :external_user_id
+  validate :verify_mal_credentials
+
+  def verify_mal_credentials
```
Make sure you check `sync_to_mal?` before this validation — you should be able to do `validate :verify_mal_credentials, if: :sync_to_mal?`

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,166 @@
+        end
+      else
+        # manga
+        if response['read_status']
```
Yo, listen to the dog!

toyhammered replied:
I actually don't know a nice way to do this... Only reason why I ignored :|

NuckChorris replied:
Instead of doing this:

```ruby
if foo
  # do stuff
else
  if bar
    # do other stuff
  else
    # do other other stuff
  end
end
```

Do this

```ruby
if foo
  # do stuff
elsif bar
  # do other stuff
else
  # do other other stuff
end
```

NuckChorris reviewed (changes requested):
Much closer, though I almost feel like putting this in `LinkedProfile` was a horrible mistake :/

Like maybe it should be... idk what to call the other model, but like, split the model backing the links on your profile from a separate model backing stuff like this and the Facebook/Twitter login?

It just seems like the logic for what *should* be a simple model (profile links) is gonna get messy really fast as we add more hookups, so if the hookups were separate from the links, it might be simpler to comprehend?

cybrox commented:
> Like maybe it should be... idk what to call the other model, but like, split the model backing the links on your profile from a separate model backing stuff like this and the Facebook/Twitter login?

I mean, we have iterated over multiple solutions for this in the dev chat before and I think we can agree, that we didn't find a solution that fits quite all needs. With the amount of linkable profiles that are backed by an actual integation (be it oauth or mal's crappy basic auth), is quite high, so separating it would introduce useless roundabouts imo, either for us or also the user, depending on which of the discussed scenarios we pick.

I think it will probably be best to keep linked profiles a single model but before we go too far with it, look at all the main sites we actually planned on integrating and flesh out which properties we need, in order to display and/or actually integrate it.

toyhammered committed:
change database table relationship for linked_* stuff

toyhammered committed:
Change LinkedProfile and LinkedSite
  - renames LinkedProfile -> LinkedAccount
  - renames LinkedSite -> ProfileLinkSite
    - Factory Girl Linter error with ProfileLinkSite
  - creates ProfileLink
  - Fixes all files related to old naming including:
    - Policies
    - Resources
  - Fixes all tests to pass with new/renamed database tables

toyhammered committed:
add STI to LinkedAccount
  - also fixes some minor issues (rubocop)

toyhammered commented on app/models/profile_link.rb:
```diff
@@ -0,0 +1,27 @@
+#
+#  id                   :integer          not null, primary key
+#  url                  :string           not null
+#  profile_link_site_id :integer          not null, indexed, indexed => [user_id]
```
go... die...

toyhammered committed:
fix rubocop errors

toyhammered committed:
fix rubocop linelength

NuckChorris commented on app/models/linked_account.rb:
```diff
@@ -0,0 +1,35 @@
+class LinkedAccount < ApplicationRecord
+  belongs_to :user, required: true
+  # encyrpt the token
+  attr_encrypted :token, key: Base64.encode64(ENV['ATTR_ENCRYPT_KEY'])
```
Shouldn't this be *decode64*?

toyhammered replied:
I _think_ what are we going to be storing in heroku (will it be an encoded or decoded key?)

NuckChorris replied:
It *will* be encoded as base64 (so we don't store binary data directly), so you need to *decode* it here

NuckChorris replied:
(P.S. how do I generate the key, is it just random data?)

toyhammered replied:
yup just random, I went into irb originally to generate the binary data.

NuckChorris commented on app/models/linked_account.rb:
```diff
@@ -0,0 +1,35 @@
+  belongs_to :user, required: true
+  # encyrpt the token
+  attr_encrypted :token, key: Base64.encode64(ENV['ATTR_ENCRYPT_KEY'])
+  # validates_presence_of :url, if: :private?
```
Can you delete this instead of commenting it out?

NuckChorris commented on app/models/linked_account/my_anime_list.rb:
```diff
@@ -0,0 +1,27 @@
+
+      return true if response.code == 200
+
+      errors.add(:token, "#{response.code}: #{response.body}")
```
This is a terrible error message, check if the error is 403 and give something reasonable like "Username or password was incorrect"

NuckChorris commented on app/models/profile_link_site.rb:
```diff
@@ -13,7 +13,7 @@
 #
 # rubocop:enable Metrics/LineLength
```
Can probably drop the `share_to`, `share_from` and `link_type` columns tbh

NuckChorris replied:
Sure, but that's part of LinkedAccount, not ProfileLink

toyhammered replied:
ahhh ok, wonder why we originally had it there then...

NuckChorris replied:
Because, remember back to like 3 days ago

When the ProfileLink and LinkedAccount were the same system

Before you, yourself, changed it

toyhammered replied:
such a weird thing we did prior to this, confused the shit outta me.

NuckChorris replied:
You and me both, man.  This new split is much easier to comprehend I think

NuckChorris commented on app/resources/profile_link_resource.rb:
```diff
@@ -0,0 +1,7 @@
+class ProfileLinkResource < BaseResource
+  immutable
```
This probably shouldn't be immutable, that would mean nobody can edit the links on their profile

toyhammered replied:
oh... I mis-thought that. I forgot that someone could add a profile link and also remove it if they don't want it showing anymore... ops.

NuckChorris commented on app/resources/linked_account_resource.rb:
```diff
@@ -0,0 +1,17 @@
+    else
+      super - [:token]
+    end
+  end
```
I think this is unnecessary, since the policy can just prevent users from seeing entire LinkedAccounts of other users

toyhammered replied:
so I should just remove the whole `fetchable_fields`? (I personally never understood the point of this, you just said copy it from somewhere lol)

NuckChorris replied:
`fetchable_fields` is a method describing which attributes are visible to a given user — so, for example, only you can see your own email even though everybody can see your `User` (they just get `email: nil`)

So when this was combined previously, we had to display the row to all users (if it wasn't private) but hide the `token` field (so people can't steal your MAL password)

Now that it's split, you can hide the entire row from other users using the policy scope, instead of just hiding the `token` field with `fetchable_fields`

toyhammered replied:
ooooh interesting. So adding the policy scope pretty much replaced the `fetchable_fields`.

NuckChorris replied:
Pretty much.  Really, I'd say what replaced it is the splitting of private (LinkedAccount) from public (ProfileLink)

NuckChorris commented on app/resources/linked_account_resource.rb:
```diff
@@ -0,0 +1,17 @@
+
+  attribute :token
+  attributes :external_user_id, :share_to, :share_from, :sync_to
+  # :type is reserved for STI
```
For ListImport we do `alias_attribute :kind, :type` in the model, and then expose `kind` as an attribute here.  Also can you combine the two lines here?  There's no reason to have 'em separate

NuckChorris commented on app/resources/linked_account_resource.rb:
```diff
@@ -0,0 +1,17 @@
+class LinkedAccountResource < BaseResource
+  model_hint model: LinkedAccount::MyAnimeList
```
I wanna say I think the class name needs to be in a string, like `"LinkedAccount::MyAnimeList"`

I'm not sure, check ListImportResource

toyhammered replied:
List import has it not as a string. Docs are the same. So the way I have it is right.

NuckChorris replied:
Aight then :+1:

NuckChorris commented on app/policies/linked_account_policy.rb:
```diff
@@ -1,12 +1,12 @@
 
   class Scope < Scope
     def resolve
       return scope.where(private: false) unless user
```
The entire scope method I think can be replaced with `scope.where(user: user)`, since you should only be able to see your own (and admins should not be allowed to peek)

toyhammered replied:
yes actually forgot to change this, we already removed private column anyways.

NuckChorris commented on app/services/my_anime_list_sync_service.rb:
```diff
@@ -0,0 +1,161 @@
+    @method = method
+  end
+
+  def execute_method
```
Probably wanna check if `mal_media` is nil here, to handle stuff that's in our anime database but not MAL's

toyhammered replied:
should i just do an `errors.add` so that we know where issues are?

NuckChorris replied:
Can't do that within a service object.  We'll worry about that later, for now just `return` early to skip the sync

NuckChorris commented on config/routes.rb:
```diff
@@ -605,6 +605,7 @@
     jsonapi_resources :roles
     jsonapi_resources :user_roles
     jsonapi_resources :reports
+    jsonapi_resources :linked_accounts
```
Can you also add one for `:profile_links`?

toyhammered commented on app/models/profile_link_site.rb:
```diff
@@ -13,7 +13,7 @@
 #
 # rubocop:enable Metrics/LineLength
```
are we not going to allow other sites to post stuff to kitsu?

toyhammered committed:
fix all errors from Nuck's review

toyhammered commented on app/policies/linked_account_policy.rb:
```diff
@@ -0,0 +1,11 @@
+
+  class Scope < Scope
+    def resolve
+      return scope.where(user: user)
```
fixed, just gonna wait on Travis to make sure there is nothing else.

toyhammered committed:
fix profile_link-sites factory and rubocop error

toyhammered committed:
remove unnecissary profile_link_site columns from model

NuckChorris commented on app/models/library_entry.rb:
```diff
@@ -134,5 +143,12 @@ def activity
     end
+
+    # Sync MAL updates if linked profile exists
+    MyAnimeListSyncWorker.perform_async(self, 'create/update') if sync_to_mal?
```
One final thing and then I think we can merge this — you don't wanna pass the whole object into a worker, it gets nasty and messy

Instead of `self`, pass `id`, and then in the worker, do `LibraryEntry.find(library_entry_id)`

toyhammered committed:
pass in library entry id rather than whole object

toyhammered committed:
add end tag that was removed during rebase.. oopsie?
